File: Gomoku-Game.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 15;
const int M = 15;
const int a_row = 5;
const int n_players = 2;
char marks[n_players] = {'X', 'O'};
char grid[N][M];
char temp[N][M];
//This function prints the grid of Gomoku as the game progresses
void print_grid() {
	for (int i = 0; i< n_players; i++) {
        cout << "Player " << i+1 << ": " << marks[i] << "  ";
        if (i < n_players-1)
            cout << "vs  ";
	}
	cout << "\n";
	cout << "--";
    for (int i = 0; i < M; cout << "---", i++);
    cout << "--\n";
    for (int i = 0; i < N; i++) {
        cout << "|  ";
        for (int j = 0; j < M; j++)
            cout << grid[i][j] << "  ";
        cout << "|\n";
        cout << "--";
        for (int i = 0; i < M; cout << "---", i++);
        cout << "--\n";
    }
}
//This function checks if there is a win row
bool check_row(char grid[N][M]){
	for(int row = 0;row<N;row++){
		for(int col = 0;col<(M-(a_row-1));col++){
			for(int i = col+1;i<col+a_row;i++){
				if(grid[row][col]==grid[row][i] && ( grid[row][col] == 'X' || grid[row][col] == 'O' ) ){
					if(i == col+(a_row-1)){
						return true;
					}
					continue;
				}
				else{
					break;
				}
			}
		}	
	}
	return false;	
}
//This function checks if there is a win column
bool check_col(char grid[N][M]){
	for(int col = 0;col<M;col++){
		for(int row = 0;row<(N-(a_row-1));row++){
			for(int i = row+1;i<row+a_row;i++){
				if(grid[row][col]==grid[i][col] && ( grid[row][col] == 'X' || grid[row][col] == 'O' ) ){
					if(i == row+(a_row-1)){
						return true;
					}
					continue;
				}
				else{
					break;
				}
			}
		}	
	}
	return false;	
}

//This function checks if there is a win diagonal
bool check_diag(char grid[N][M]){
	for(int row = 0; row < (N-(a_row-1)); row++){
		for(int col = 0; col < (M-(a_row-1)); col++){

			for( int x=row+1 , y=col+1; x < (row+a_row) , y < (col+a_row); x++ , y++ ){
				
				if( grid[row][col]==grid[x][y] && (grid[row][col]=='X'|| grid[row][col]=='O') ){
					if( x == (row+(a_row-1)) ){
						return true;
					}
					continue;				
				}
				else{
					break;
				}
				
			}
			
		}
	}
	for(int row = N-1; row > (N-(N-a_row)); row--){
		for(int col = 0; col < (M-(a_row-1)) ; col++){
			for(int x = row-1 , y = col+1; x < (row-a_row) , y < (col+a_row); x-- , y++){
				if(grid[row][col]==grid[x][y] &&(grid[row][col]=='X'||grid[row][col]=='O')){
					if(x == (row-(a_row-1)) ){
						return true;
					}
					continue;				
							
				}
				else{
					break;
				}
				
			}
			
				
			
		}
	}
	return false;
	
}

//This function checks if the game has a win state or not
bool check_win() {
	if(check_row(grid) || check_col(grid) || check_diag(grid)){
		return true;
	}
	return false;
}
//This function copy the main array(grid) to a temporary array(temp) and replace empty spaces with marks
void arr_try(char mark){
	for(int i = 0;i<N;i++){
		for(int j = 0;j<M;j++){
			if(grid[i][j]==' '){
				temp[i][j] = mark;
			}else{
				temp[i][j] = grid[i][j];
			}
		}		
	}
}
//This function checks if the game has a tie state or not for the given mark
bool check_tie_player(char mark) {
	arr_try(mark);
	if(check_row(temp)||check_col(temp)||check_diag(temp)){
		return false;
	}
	return true;
}

//This function checks if the game has a tie state or not
bool check_tie() {
    bool all_tie = true;
    for (int i = 0; i < n_players; i++)
        if (!check_tie_player(marks[i]))
            all_tie = false;
    return all_tie;
}

//This function checks if given cell is empty or not 
bool check_empty(int i, int j) {
	if(grid[i][j]==' '){
		return true;
	}
	else{
		return false;
	}
}
//This function checks if given position is valid or not 
bool check_valid_position(int i, int j) {
	if(( i >= 0 && i < N ) && ( j >= 0 && j < M )){
		return true;
	}
	else{
		return false;
	}
}
//This function sets the given mark to the given column
void set_cell(int i, int j, char mark) {
	grid[i][j] = mark;
}
//This function clears the game structures
void grid_clear() {
	for(int i = 0 ; i<N ; i++){
		for(int j = 0 ; j<M ; j++){
			grid[i][j] = ' ';
		}
	}
}
//This function reads a valid position input
void read_input(int &i, int &j) {
    cout << "Enter the row index and column index: ";
    cin >> i >> j;
    while (!check_valid_position(i, j) || !check_empty(i, j)) {
        cout << "Enter a valid row index and a valid column index: ";
        cin >> i >> j;
    }
}
//MAIN FUNCTION
void play_game() {
    cout << "Gomoku Game!\n";
    cout << "Welcome...\n";
    cout << "============================\n";
    int player = 0;
    while (true) {
        //Prints the grid
        print_grid();
        //Read an input from the player
        cout << "Player " << marks[player] << " is playing now\n";
        int i, j;
		read_input(i, j);
        //Set the input position with the mark
        set_cell(i, j, marks[player]);
        //Check if the state of the grid has a win state
        if (check_win()) {
            //Prints the grid
            print_grid();
            //Announcement of the final statement
            cout << "Congrats, Player " << marks[player] << " is won!\n";
            break;
        }
        //Check if the state of the grid has a tie state
        if (check_tie()) {
            //Prints the grid
            print_grid();
            //Announcement of the final statement
            cout << "Woah! That's a tie!\n";
            break;
        }
        //Player number changes after each turn
        player = (player + 1) % n_players;
    }
}
int main() {
    while (true) {
    	grid_clear();
    	play_game();
    	char c;
    	cout << "Play Again [Y/N] ";
    	cin >> c;
    	if (c != 'y' && c != 'Y')
    		break;
    }
}

File: Gomoku-Game.py
N, M = 15, 15
a_row = 5
n_players = 2
marks = ['X', 'O']
grid = []

#This function prints the grid of Gomoku as the game progresses
def print_grid():
    for i in range(n_players):
        print('Player %d: %c  ' % (i+1, marks[i]), end='')
        if i < n_players-1:
            print('vs  ', end='')
    print()
    print('--' + '---' * M + '--')
    for i in range(N):
        print(end='|  ')
        for j in range(M):
            print(grid[i][j], end='  ')
        print(end='|')
        print()
        print('--' + '---' * M + '--')
#This function checks the existence of any winning row
def check_rows(grid):
    for row in range(N):
        for col in range(M-(a_row-1)):
            for i in range(col+1,col+a_row):
                if grid[row][col]==grid[row][i] and grid[row][col] in marks :
                    continue
                else:
                    break
            else:
                return True
    return False         
#This function checks the existence of any winning column
def check_cols(grid):
    for col in range(M):
        for row in range(N-(a_row-1)):
            for i in range(row+1,row+a_row):
                if grid[row][col]==grid[i][col] and grid[row][col] in marks :
                    continue
                else:
                    break
            else:
                return True
    return False 
#This function checks the existence of any winning diagonal
def check_diags(grid):
    for row in range(N-(a_row-1)):
        for col in range(M-(a_row-1)):
            j = col + 1
            for i in range(row+1,row+a_row):
                if grid[row][col] == grid[i][j] and grid[row][col] in marks :
                    j+=1
                    continue
                else:
                    break
            else:
                return True

    for row in range(N-1,N-(N-a_row),-1):
        for col in range(M-(a_row-1)):
            j = col + 1
            for i in range(row-1,row-a_row,-1):
                if grid[row][col] == grid[i][j] and grid[row][col] in marks :
                    j+=1
                    continue
                else:
                    break
            else:
                return True
    return False

#This function checks if the game has a win state or not
def check_win():
    if check_rows(grid) or check_cols(grid) or check_diags(grid):
        return True
    else:
        return False
#This function makes a temporary copy from grid and fill the spaces with specific mark
def list_try(mark):
    temp=[]
    for row in range(N):
        l = []
        for col in range(M):
            if grid[row][col] == " ":
                l.append(mark)
            else:
                l.append(grid[row][col])
        temp.append(l)
    return temp
#This function checks if the game has a tie state or not for the given mark
def check_tie_player(mark):
    temp = list_try(mark)
    if check_rows(temp) or check_cols(temp) or check_diags(temp):
        return False
    else:
        return True

#This function checks if the game has a tie state or not
def check_tie():
    all_tie = True
    for mark in marks:
        if not check_tie_player(mark):
            all_tie = False
    return all_tie
	
#This function checks if given cell is empty or not 
def check_empty(i, j):
    if grid[i][j] == " ":
        return True
    else : 
        return False

#This function checks if given position is valid or not 
def check_valid_position(i, j):
    if 0<=i<N and 0<=j<M :
        return True
    else:
        return False

#This function sets the given mark to the given cell
def set_cell(i, j, mark):
    grid[i][j] = mark

#This function clears the game structures
def grid_clear():
    grid.clear()
    for i in range(N):
        l = [" "]*M
        grid.append(l)

#This function reads a valid position input
def read_input():
    i, j = map(int, input('Enter the row index and column index: ').split())
    while not check_valid_position(i, j) or not check_empty(i, j):
        i, j = map(int, input('Enter a valid row index and a valid column index: ').split())
    return i, j


#MAIN FUNCTION
def play_game():
    print("Gomoku Game!")
    print("Welcome...")
    print("============================")
    player = 0
    while True:
        #Prints the grid
        print_grid()
        #Read an input position from the player
        print('Player %s is playing now' % marks[player])
        i, j = read_input()
        #Set the player mark in the input position
        set_cell(i, j, marks[player])
        #Check if the grid has a win state
        if check_win():
            #Prints the grid
            print_grid()
            #Announcement of the final statement
            print('Congrats, Player %s is won!' % marks[player])
            break
        #Check if the grid has a tie state
        if check_tie():
            #Prints the grid
            print_grid()
            #Announcement of the final statement
            print("Woah! That's a tie!")
            break		
        #Player number changes after each turn
        player = (player + 1) % n_players


while True:
	grid_clear()
	play_game()
	c = input('Play Again [Y/N] ')
	if c not in 'yY':
		break
